src/lib/cat_stories/index.svelte

```diff
@@ -49,45 +55,194 @@
+    return largest;
+  }
+
+  function extractNameAndIndexFromId(carouselItem: HTMLElement) {
+    return carouselItem.id.split('-');
+  }
+
+  function getIdFromNameAndIndex(name: string, index: number) {
+    return `${name}-${index}`;
+  }
+
+  function getNeighboringNode(parent: HTMLElement, child: HTMLElement, next: boolean) {
+    const [childName, childIndex] = extractNameAndIndexFromId(child);
+    const childIdToFind = getIdFromNameAndIndex(childName, (childIndex + (next ? 1 : 2)) % 3);
+    return document.getElementById(childIdToFind);
+  }
+
+  function getDistanceBetweenNodes(node1: HTMLElement, node2: HTMLElement) {
+    const rect1 = node1.getBoundingClientRect();
+    const rect2 = node2.getBoundingClientRect();
+    const center1 = rect1.x + rect1.width / 2;
+    const center2 = rect2.x + rect2.width / 2;
+    return center1 - center2;
+  }
+
+  function scrollStoriesContainer(e: Event) {
+    if (!e.target) return;
+    let { scrollWidth, clientWidth, scrollLeft, childNodes } = e.target as HTMLDivElement;
+    const center = clientWidth / 2;
+    const currTimestampMS = new Date().getTime();
+
+    const scrollSpeed = scrollLeft - lastScrollLeft;
+    if (Math.abs(scrollSpeed) < 4 && currTimestampMS - lastScrollTimestampMS >= 1000) {
+      const largestChild = getLargestChild(e.target);
+      if (clientWidth + scrollLeft + 200 > (2 / 3) * scrollWidth) {
+        // At far right of scoll container, scrolling right
+        lastScrollTimestampMS = currTimestampMS;
+        clearTimeout(snapTimeout);
+
+        let xToScrollTo =
+          scrollLeft +
+          getDistanceBetweenNodes(getNeighboringNode(e.target, largestChild, false), largestChild);
+        if (xToScrollTo > scrollWidth) {
+          xToScrollTo -= scrollWidth;
+        }
+        if (xToScrollTo < 0) {
+          xToScrollTo += scrollWidth;
+        }
+        snapTimeout = setTimeout(() => {
+          scrollIfNotMoving(() => (e.target as HTMLDivElement).scrollTo({ left: xToScrollTo }));
+        }, 50);
+      } else if (scrollLeft < 400) {
+        // At far left of scroll container, scrolling left
+        lastScrollTimestampMS = currTimestampMS;
+        clearTimeout(snapTimeout);
+        let xToScrollTo =
+          scrollLeft +
+          getDistanceBetweenNodes(getNeighboringNode(e.target, largestChild, true), largestChild);
+        if (xToScrollTo > scrollWidth) {
+          xToScrollTo -= scrollWidth;
+        }
+        if (xToScrollTo < 0) {
+          xToScrollTo += scrollWidth;
+        }
+        snapTimeout = setTimeout(() => {
+          scrollIfNotMoving(() => (e.target as HTMLDivElement).scrollTo({ left: xToScrollTo }));
+        }, 50);
+      }
+    }
+
+    scaleChildren(childNodes, center);
+
+    lastScrollLeft = scrollLeft;
+  }
+
+  onMount(() => {
+    const storiesContainer = document.querySelector('.stories-container');
+    if (storiesContainer) {
+      storiesContainer.addEventListener('scroll', scrollStoriesContainer);
```

Comment:
I should've mentioned this earlier, but a big performance improvement comes from doing this:

`storiesContainer.addEventListener('scroll', scrollStoriesContainer, { passive: true });`

It tells the browser we won't hijack the normal scroll process by calling `preventDefault` (which means it can do a lot of performance optimizations).